src/qibo/quantum_info/quantum_networks.py

```diff
@@ -44,161 +45,204 @@ class QuantumNetwork:
+        tensor,
+        partition: Optional[Union[List[int], Tuple[int]]] = None,
+        system_input: Optional[Union[List[bool], Tuple[bool]]] = None,
         pure: bool = False,
         backend=None,
     ):
-        self._run_checks(partition, system_output, pure)
+        self._run_checks(partition, system_input, pure)
 
-        self._matrix = matrix
+        self._tensor = tensor
         self.partition = partition
-        self.system_output = system_output
+        self.system_input = system_input
         self._pure = pure
         self._backend = backend
-        self.dims = reduce(mul, self.partition)
 
-        self._set_tensor_and_parameters()
+        self._set_parameters()
 
-    def matrix(self, backend=None):
-        """Returns the Choi operator of the quantum network in matrix form.
+        if len(self.partition) > 0:
+            self.dims = reduce(mul, self.partition)
+        else:
+            self.dims = 1
 
-        Args:
-            backend (:class:`qibo.backends.abstract.Backend`, optional): Backend to be used
-                to return the Choi operator. If ``None``, defaults to the backend defined
-                when initializing the :class:`qibo.quantum_info.quantum_networks.QuantumNetwork`
-                object. Defaults to ``None``.
+    @staticmethod
+    def _order_tensor2operator(n: int):
+        order = list(range(0, n * 2, 2)) + list(range(1, n * 2, 2))
+        return order
 
-        Returns:
-            ndarray: Choi matrix of the quantum network.
-        """
-        if backend is None:  # pragma: no cover
-            backend = self._backend
+    @staticmethod
+    def _order_operator2tensor(n: int):
+        order = list(sum(zip(list(range(0, n)), list(range(n, n * 2))), ()))
+        return order
 
-        return backend.cast(self._matrix, dtype=self._matrix.dtype)
+    @classmethod
+    def _operator2tensor(cls, operator, partition: List[int]):
+        n = len(partition)
+        order = cls._order_operator2tensor(n)
 
-    def is_pure(self):
-        """Returns bool indicading if the Choi operator of the network is pure."""
-        return self._pure
+        try:
+            tensor = np.transpose(operator.reshape(list(partition) * 2), order)
+            return tensor.reshape([dim**2 for dim in partition])
+        except:
+            raise_error(
+                ValueError,
+                "``partition`` does not match the shape of the input matrix. "
+                + f"Cannot reshape matrix of size {operator.shape} to partition {partition}",
+            )
 
-    def is_hermitian(
-        self, order: Optional[Union[int, str]] = None, precision_tol: float = 1e-8
+    @classmethod
+    def from_operator(
+        cls,
+        arr: np.ndarray,
+        partition: Optional[Union[List[int], Tuple[int]]] = None,
+        system_input: Optional[Union[List[bool], Tuple[bool]]] = None,
+        pure: bool = False,
+        backend=None,
     ):
```

Comment:
This method receives a variable named `arr` being typed as `numpy.ndarray`. Later, this method is overwritten by other classes, with this same variable being renamed as `tensor`. We should stick to one name throughout. Moreover, as we discussed before, this can be more than just a `numpy.ndarray`, so I'd just leave it untyped for simplicity.